src/karapace/schema_models.py

```diff
@@ -207,18 +242,33 @@ def parse(
 ) -> ParsedTypedSchema:
     if schema_type not in [SchemaType.AVRO, SchemaType.JSONSCHEMA, SchemaType.PROTOBUF]:
         raise InvalidSchema(f"Unknown parser {schema_type} for {schema_str}")
-
     parsed_schema: Draft7Validator | AvroSchema | ProtobufSchema
     if schema_type is SchemaType.AVRO:
         try:
+            if dependencies:
+                schemas_list = AvroResolver(schema_str, dependencies).resolve()
+                names = AvroNames(validate_names=validate_avro_names)
+                merged_schema = None
+                for schema in schemas_list:
+                    # Merge dep with all previously merged ones
+                    merged_schema = make_avsc_object(json.loads(schema), names)
+                merged_schema_str = str(merged_schema)
+            else:
+                merged_schema_str = schema_str
             parsed_schema = parse_avro_schema_definition(
```

Comment:
I have realized now that it is possible to pass a list of schema strings to `avro.schema.parse()`:
https://github.com/aiven/avro/blob/5a82d57f2a650fd87c819a30e433f1abb2c76ca2/lang/py/avro/schema.py#L1192

We currently use `parse()` to parse a single schema (see `parse_avro_schema_definition()`).

When called with a list of schemas it returns a `UnionSchema` (through `make_avsc_object()`), which itself contains a list of parsed schemas (`UnionSchema.schemas`).
The construct of `UnionSchema` seems to be doing what we want (loop on the schemas, call `make_avsc_object()` and remember `names`).

I think passing to `parse()` the reference schemas + the main schema (the main schema should be passed as last in the list) and then getting the last schema from `UnionSchema.schemas` should be equivalent to what's the PR now.

If it works then we can delegate that logic to the Avro library